### DISCUSSION TITLE
resolve windows environment variables for paths

### DISCUSSION THREAD

**Discussion by goyalyashpal** [msys2/msys2-runtime#147] [Q&A] [answered]
will it make sense to resolve the windows environment variables (enclosed in %) for paths in the msys2 shell?

this will make the windows environment variables like `%localappdata%` and others to work.
these are supported in windows stuff like windows explorer, `cmd`  (but surprisingly, not in `powershell`)

i was thinking of using like this: 

```console
$ cd %localappdata%
```

```console
$ mv %userprofile%/some-file /d/my-dir
```

currently these variables are not resolved and error is shown: `bash: cd: %localappdata%: No such file or directory`

**Comment by Biswa96** [comment]
> will it make sense to resolve the windows environment variables (enclosed in %) for paths in the msys2 shell?

No, it's a convention to use `%` for environment variables in Windows or DOS like OS. Whereas `$` is used for unix-like environment.

In case of bash, `%` is used for other purposes, like string format specifier, pattern matching, arithmetic operation etc. More info can be found in bash man page https://www.man7.org/linux/man-pages/man1/bash.1.html.

> this will make the windows environment variables like %localappdata% and others to work.

If you want to use cmd like syntax you can use msys2 by adding the installation directories in %PATH% env. variable in cmd or powershell. But some tools may not work as expected.

**Comment by goyalyashpal** [comment]
> cmd like syntax [with msys2 tools], you can use msys2 by adding the installation directories in %PATH% env. variable in cmd or powershell

ahw, really nice. yep, `ls "%localappdata%"` worked from `cmd` (not from `powershell` - this might be due to old powershell instance, but couldn't figure out its version lol)

also, `powershell /?` doesn't work from `bash` - so, its interesting that the environment interacts and modifies the command invocation. i had only read this in msys2 documentation, but experiencing it this drastically for the first time.

kinda amazing.

also, i was thinking of adding these windows environment variables in the `$` format in one of my msys config file, then doing something like `ls "${localappdata}"`

**Comment by Biswa96** [comment, marked as answer]
> ls "%localappdata%" worked from cmd

It works from msys2 bash also. See the following output.

```
$ echo $LOCALAPPDATA
C:\Users\x\AppData\Local

$ ls -lash $LOCALAPPDATA | wc -l
42
```

**Reply by realyukii** [reply]
Did you set it manually or it's somehow inherited? in my case, the tmp is pointing to /tmp instead of `C:\Users\USER\AppData\Local\Temp`
```
USER@LAPTOP MSYS ~
# echo $TMP
/tmp
USER@LAPTOP MSYS ~
# echo $TEMP
/tmp
USER@LAPTOP MSYS ~
# echo $LOCALAPPDATA
C:\Users\USER\AppData\Local
USER@LAPTOP MSYS ~
#
```

I'm wondering if there's an option to inherit $TMP or $TEMP as well.

**Reply by goyalyashpal** [reply]
admin / elevated shell?

**Reply by realyukii** [reply]
> admin / elevated shell?

not sure, the snippet above is from msys2's ssh, I guess it's admin (since `#` mark).

**Comment by goyalyashpal** [comment]
> [with the dollar instead of percents] It works from msys2 bash also. See the following output.

brilliant :) super neat, one more reason to love msys2 from my side

reminder to self, 
- yes, it has to be case sensitive
- any wrong variable with result in empty value, which will lead to unexpected results (will likely involve the `~` dir), so, try to figure a way to output the executed commands (like the `debug` or smth)

**Comment by hmartinez82** [comment]
In powershell, it's `$env:<name of variable>` . E.g.: `$env:localappdata`

**Reply by goyalyashpal** [reply]
- wow, it resolves it even while typing for tab completion, nice :)
- other differences: it shows error for non defined i.e. null variables (for e.g. `userprofiles`) - which is better than silently proceeding to completion. this silent completion in bash causes wrong behaviour in some cases

- one more thing which improves the experience in powershell: ability to undo/redo the typing - this is probably the biggest thing i miss in bash
- last thing missing in bash would be colored syntax highlighting of the input commands

**Reply by hmartinez82** [reply]
Use powershell from Windows Terminal, it has color support

**Reply by goyalyashpal** [reply]
yep, i was saying about it missing in bash

**Reply by sskras** [reply]
> yep, i was saying about it missing in bash

... in `libreadline`, I guess.

EDIT: And it seems it supports the Undo: [Readline Cheat Sheet](https://readline.kablamo.org/emacs.html#:~:text=x%20Ctrl%2Du-,Incremental%20undo,-%2C%20separately%20remembered%20for)

It's the `Ctrl-_` combo (`Ctrl`+`Shift`+`-` on my keyboard).

Sadly, Redo doesn't seem to be supported.
